wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+These two naive approaches are unsatisfactory because they are missing a few critical elements:
+
+1. They did not use randomness.
+2. Their position mapping function for each leaf is invertible.
+3. They did not expand the search space for brute-force attacker (confined within M<sup>2</sup>)
+
+To address these issues, we introduce the following technique:
+
+1. For each leaf, we use the corresponding OTP as the seed for a random number generator, to generate 4 consecutive numbers between [1, 4M]. Note this mapping is deterministic (hence repeatable) but is not invertible.
+2. We break down the leaf into 4 parts of 8-byte chunks, and store the chunks at the position given by the 4 random numbers generated in the previous step.
+3. If any two chunks fall into the same position, they are stored in a list at that position. For brevity we call this scenario a "collision".
+4. For each pair of neighbors, we swap the position of their chunks since we only need the neighbor to complete the Merkle Proof, not the leaf itself.
+
+At runtime, the client can easily recover the positions of each chunk by repeating (1). If there is a list consisting of more than one element at any position, the client can simply enumerate all possible combinations using each of the element in the list. In practice, the chance that any position has more than 1 or 2 elements in the list is very slim, since our the size of our random space in (1) equals to the number of elements we want to generate. We can further increase the space to increase sparsity and reduce collision. After recovering the chunks, the client can complete the Merkle Proof and proceed as usual. If there are multiple combinations of chunks due to collision, the client may test for each combination of whether it is correct, by computing the current leaf using **GenLeaf** and the user provided OTP, then check whether the hash of the concatenation of the current leaf and the neighbor chunk-combination candidate equals to the required hash value in the parent node corresponding to current timestamp or index.
+
+With this approach the number of operations required for simple brute-force is increased to (4M)<sup>8</sup> since the attacker would have to enumerate all possible combinations of chunks for each pair of neighbors and test whether they produce the required hash value in a any parent node. There may be more efficient brute-force attacks which we have not explored or analyzed.
+
+Overall, this techniques is the least straightforward, but may be the most secure and the fastest for runtime execution. In any case, this technique can be composed with the techniques in Part I or II or both to obtain multiplicative security protection, since they operate at different part of the process.
```

Comment:
From @ivan-homoliak-sutd : 

> You do not assume intermediary nodes of Merkle tree already submitted to blockchain within some old proofs. The attacker is able to permutate, I'd guess, up to 4 or 8 tuples of EOTPs from the client's memory and check it against some future intermediary node (at the level 16 or 17 from the root) submitted in some Mrekle proof. So, in theory, the attacker has 8*30s = 240sec to find the correct permutation (assuming that he can do max. 8 touples, maybe she can do more and will have more time).

Reply:
Denote the leaves as nodes at level 0. It is assumed that correct, unscrambled parent nodes (level 1) are visible to the attacker at the client side. Previous leaves submitted to the blockchain are also visible to the attacker. However, to find the correct permutation of chunks for the current leaf, the attacker still needs to enumerate over a large space

Reply:
OK, it is seems fine. Thanks.